fix(platform-browser): insert APP_ID in styles, contentAttr and hostAttr

petersalomonsen opened the pull request:
## PR Checklist
Does please check if your PR fulfills the following requirements:

- [x ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ x] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: 16676


## What is the new behavior?

APP_ID is inserted between nghost/ngcontent and component id in styles, contentAttr and hostAttr as it is supposed to do according to the documentation (https://angular.io/guide/component-styles)
## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Existing tests have been modified to check for the APP_ID presence. Documentation already states that this should be present, so the change is bringing the implementation closer to how the documentation says it should work.

petersalomonsen committed:
fix(platform-browser): insert APP_ID in styles, contentAttr and hostAttr

splincode commented:
@petersalomonsen what is the status of your pull request? is it still relevant? please fix conflicts if they exist

petersalomonsen commented:
@splincode It is still relevant. Without it you will have overlapping styles when having multiple angular apps on the same page, which is caused by the app id no being included as it is supposed to be according to the documentation here: https://angular.io/guide/component-styles#inspecting-generated-css

glipecki commented:
Additionally, current behavior is misleading as APP_ID documentation states that APP_ID should be used as a part of generated attributes for ViewEncapsulation.Emulated (https://angular.io/api/core/APP_ID).

This PR would help me a lot! :-)

glipecki commented:
@petersalomonsen will you fix pr checks?

glipecki commented:
@petersalomonsen Are you going to fix PR checks?

petersalomonsen commented:
@glipecki Can have a look at it.

petersalomonsen commented:
`packages/core/test/render3/imported_renderer2.ts` was added after my initial PR. The DomRenderer2Factory was instanced without the app id parameter (that I added in my PR), so I added a dummy app id to get around it. Tests passed locally so hopefully passes the CI too? (I'm not 100% up to speed on all that's going on in the CI tests).

splincode commented:
@petersalomonsen This pull request is big!

petersalomonsen commented:
@splincode Did I merge / rebase the wrong way?

I did this:

```git rebase master -i
git push -f````

(as written here: https://github.com/angular/angular/blob/master/CONTRIBUTING.md - if changes are suggested (10))

I've only made two commits.

petersalomonsen commented:
@splincode Should never have used the sync button in visual studio code that created a merge commit. Merge commit is removed now.

splincode commented:
@petersalomonsen Can you edit the code according to the formatting rules (default tslint.json)?

petersalomonsen commented:
@splincode Now combined into one commit. And I've also run `gulp format`.  However `gulp lint` fails on validating another commit message (`fix(tsc-wrapped): emit exports metadata in flat modules (#17893)`) - so not able to verify my changes.

petersalomonsen commented:
@splincode Tests are passing locally and in circleci now (I had to change my ulimits as mentioned here https://github.com/angular/angular/issues/21020 in order to see the same failures). However I cannot see the travis-ci failure locally. Any ideas?

petersalomonsen commented:
All CI tests are passing now. What more can I do for this PR to be accepted?

aaaa0441 commented:
@petersalomonsen this PR looks great! I also need this fix to allow our micro-apps to have non-conflicting styles when using https://github.com/CanopyTax/single-spa.

Do you know what's preventing this PR from being approved/merged? (Sorry that I'm not familiar with the contribution process.)

aaaa0441 commented:
@petersalomonsen Thanks for responding. I got your changes and applied the diff to my local `node_modules/@angular/platform-browser` dependency using `patch-package`. Works well.

I definitely prefer Angular to have this fixed instead of monkey patching it the way I am doing.  However, I have no other way for our project to move on now. =(

Really hope this gets merged soon. Who can we @ to get this expedited?

petersalomonsen commented:
@aaaa0441 I think @chuckjaz is member of the Angular team and maybe can give some info on what needs to be done to get the merge through.

aaaa0441 commented:
Hi @chuckjaz, could you please share how this PR can proceed to get reviewed, approved & merged?

me-12 commented:
I am also dependent on this PR to get multiple angular apps running on the same page. @petersalomonsen  is there any chance that you could resolve this conflicts in the near term?

tnicola commented:
@IgorMinar @petersalomonsen Any news on this?Still blocked from more than one year!This represent a huge CSS isolation problem for multiple Angular apps and/or @angular/elements.

aaaa0441 commented:
Hi @petersalomonsen, I have been cherry-picking your commit for the version of Angular that we use (the latest being 6.1.6). From a certain version, there needs to be one additional line related to dependency injection of the `APP_ID` token. I added it in my fork: https://github.com/aaaa0441/angular/commit/8b55dabb30e777e69d65a903dadfd53e7d4c5c30#diff-3.

Without this line change, at least on Angular 6.1.6, the generated component attribute becomes something like `_ngcontent-undefined-c1`.

petersalomonsen commented:
Thanks @aaaa0441 - I've added that to my PR now.

petersalomonsen commented:
I can see the ivy renderer tests are failing, and that's of course since they don't expect the app-id in the component attribute. Probably an easy fix, but is it worth looking into? Will this PR be merged at all?

aaaa0441 commented:
No worries, mate @petersalomonsen. I also badly want this get merged in Angular. Otherwise, we'll have to continue building our own version with this commit cherry-picked.

rahulrsingh09 commented:
Any Leads to when this is going to get merged, A lot is resting on this feature @angular-jira-bot

rsibanez89 commented:
Hi @petersalomonsen, can you please have a look at this PR and try to get it merged? Thanks.

smasala commented:
@petersalomonsen is there an update in regards to merging this PR?

Lars-Wunderlich commented:
This is actually really important for us to setup a structure of micro-app. Otherwise Angular in its current state is not useable in the enterprise.

petersalomonsen commented:
OK - I'll try aligning my branch with the latest Angular again, and hopefully they'll merge my changes. Been in sync and tests passing quite some times now, but the PR has not been picked up by the Angular team.

smasala commented:
@petersalomonsen that would be great. Here's hoping.

petersalomonsen commented:
Rebased with latest upstream. Only the IVY test failing now

smasala commented:
@petersalomonsen thanks! @filipesilva what's the status here? Thanks

filipesilva commented:
@smasala I'm afraid I'm not the right person to ask really. I mostly take care of the Angular CLI repository, not this one. @mhevery and @kara are probably the right people to review this PR.

smasala commented:
> @smasala I'm afraid I'm not the right person to ask really. I mostly take care of the Angular CLI repository, not this one. @mhevery and @kara are probably the right people to review this PR.

Thanks

mleibman commented:
An easier fix might be to use a global (window-scope) variable for the auto-incremented component ID instead of a module-level one:

https://github.com/angular/angular/blob/885f1af509eb7d9ee049349a2fe5565282fbfefb/packages/core/src/view/util.ts#L59-L77

petersalomonsen commented:
Problem with an auto-increment component id is that it would not be according to the documentation: https://angular.io/guide/component-styles#inspecting-generated-css

As said in the beginning this pull request is just about making it work like the documentation says it should.

bogdan-kuternoga commented:
@petersalomonsen I think you have to make some party when this PR will be merged xD

@splincode @mhevery @kara 
But if seriously, such approach to code review discourages.

I don't want to tell you how to prioritize issues, I just can tell you how it looks from user perspective :) . And as for me something  what is documented and doesn't work in 100% cases should be fixed ASAP instead of waiting for 2 years...

mhevery commented:
http://test/OCL:242559681:BASE:242559720:1554765423028:7c055229

mleibman commented:
@petersalomonsen looks like this is failing some unit tests, including a regex check for the _nghost attribute in https://github.com/angular/angular/blob/1102b02406ca9fac7cfc7059bc61b4d40d3dd378/packages/core/test/render3/providers_spec.ts#L1106

petersalomonsen commented:
@mleibman I believe the regex checks are fixed now. The circle_ci integration test passed on the previous push, and there was no other changes than a few spaces to please the linter (sorry for not sorting this out locally before pushing).

mleibman commented:
@petersalomonsen Thanks! I'm not an Angular team member, so I can't review/merge it myself. I'm just facilitating since I depend on this fix :)

mhevery commented:
CARETAKER: The build failure is bazel build time, which is unrelated to this PR, please merge manually.

fazil1987 commented:
is this PR merged to master ?

petersalomonsen commented:
Yes

2019-07-04 14:34 GMT+02:00 fazil1987<notifications@github.com>:
> is this PR merged to master ? 
> 
> 
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/angular/angular/pull/17745#issuecomment-508465981
